src/droid-vibrator.c

```diff
@@ -64,83 +249,131 @@ droid_vibrator_sink_can_handle (NSinkInterface *iface, NRequest *request)
 static int
 droid_vibrator_sink_prepare (NSinkInterface *iface, NRequest *request)
 {
+    DroidVibratorData *data;
+    DroidVibratorEffect *effect;
+    const NProplist *props;
+    const gchar *key;
+
     N_DEBUG (LOG_CAT "sink prepare");
 
-    DroidVibratorData *data = g_slice_new0 (DroidVibratorData);
+    props = n_request_get_properties (request);
 
-    data->request    = request;
-    data->iface      = iface;
-    data->timeout_id = 0;
+    key = n_proplist_get_string(props, HAPTIC_TYPE_KEY);
+    if (key == NULL) {
+        N_DEBUG (LOG_CAT "no effect key found for this effect");
+        return FALSE;
+    }
+
+    if (!(effect = g_hash_table_lookup (plugin_effects, key))) {
+        N_DEBUG (LOG_CAT "no effect with key %s found for this effect", key);
+        return FALSE;
+    }
+
+    data = g_slice_new0 (DroidVibratorData);
+    data->request        = request;
+    data->iface          = iface;
+    data->sequence_id    = 0;
+    data->current_effect = effect;
+    data->current_step   = effect->steps;
```

Comment:
I would have assumed current_step to go from 0 to number of steps i.e. should this be initialized to zero rather than effect->steps?